full_storage_utilization_test.py

```diff
@@ -0,0 +1,233 @@
+        self.log.info(f"Adding a node finished with time: {duration}")
+
+    def scale_in(self):        
+        self.start_throttle_rw()
+        self.log.info("Started removing a node")
+        start_time = time.time()
+        self.remove_node()
+        duration = time.time() - start_time
+        self.log.info(f"Removing a node finished with time: {duration}")
+
+    def drop_data(self, keyspace_name):
+        '''
+        Drop keyspace and clear snapshots.
+        '''
+        node = self.db_cluster.nodes[0]
+        self.log.info("Dropping some data")
+        query = f"DROP KEYSPACE {keyspace_name}"
+        with self.db_cluster.cql_connection_patient(node) as session:
+           session.execute(query)
+           #node.run_nodetool(f"clearsnapshot")
+        self.log_disk_usage()     
+
+    def perform_scale_in(self):
+        '''
+        If we are already at 90% disk utilization, first scale-out then
+        drop 20% of data to make space for scale-in operation.
+
+        If we are at 67% disk utilization, scale-in without scale-out or
+        dropping data.
+        '''
+        if self.hardlimit == 90:
+           self.scale_out()           
+           '''
+           Before removing a node, we should make sure
+           other nodes has enough space so that they
+           can accommodate data from the removed node.
+           '''
+           # Remove 20% of data from the cluster.
+           self.drop_data("keyspace_large1")
+           self.drop_data("keyspace_large2")
+           self.scale_in()
+        elif self.hardlimit == 67:
+           self.scale_in()
+
+    def perform_action(self):
+        self.log_disk_usage()
+        # Trigger specific action
+        if self.scaling_action_type == ScalingActionType.SCALE_OUT.value:
+           self.scale_out()
+        elif self.scaling_action_type == ScalingActionType.SCALE_IN.value:
+           self.perform_scale_in()
+        else:
+           self.log.info(f"Invalid ActionType {self.scaling_action_type}")
+        self.log_disk_usage()     
+
+    def test_storage_utilization(self):
+        """
+        Write data until 90% disk usage is reached.
+        Sleep for 60 minutes.
+        Perform specific action.
+        """
+        self.run_stress(self.softlimit, sleep_time=self.sleep_time_fill_disk)
+        self.run_stress(self.hardlimit, sleep_time=self.sleep_time_fill_disk)
+        self.perform_action()        
+
+    def run_stress(self, target_usage, sleep_time=600):
+        target_used_size = self.calculate_target_used_size(target_usage)
+        self.run_stress_until_target(target_used_size, target_usage)
+
+        self.log_disk_usage()
+        self.log.info(f"Wait for {sleep_time} seconds")
+        time.sleep(sleep_time)  
+        self.log_disk_usage()
+
+    def run_stress_until_target(self, target_used_size, target_usage):
+        current_usage, current_used = self.get_max_disk_usage()
+        num = 0
```

Comment:
I don't think this affects anything, but because the function is called twice in sequence and num resets inside, on the second run (hard limit), it will write data in existing keyspaces, not new ones.

Also, large keyspaces start at number 1, but small keyspaces wil start at a later number.

Reply:
ah, right. Let me fix it and ensure num doesn't reset during second run. Will also update the code to handle small keyspace properly.